File: backend/app/routes/audio.py
"""Audio API routes for ElevenLabs integration."""

import logging
from typing import Annotated

from fastapi import (
    APIRouter,
    Body,
    Depends,
    File,
    Form,
    HTTPException,
    Query,
    UploadFile,
)
from fastapi.responses import Response, StreamingResponse

from app.middleware.rate_limit import rate_limit
from app.models.schemas import SearchResult
from app.services.audio import audio_service
from app.services.auth import get_current_user

logger = logging.getLogger("nebula.audio")

router = APIRouter(prefix="/api/v1/audio", tags=["Audio"])


@router.get("/health")
async def audio_health():
    """Check if audio service is available."""
    return {
        "available": audio_service.is_available(),
        "service": "ElevenLabs",
        "features": [
            "text-to-speech",
            "speech-to-text",
            "sound-effects",
            "search-narration",
            "notifications"
        ]
    }


@router.get("/voices")
async def list_voices(
    email: str = Depends(get_current_user)
):
    """List available voices for text-to-speech."""
    voices = await audio_service.list_voices()
    return {"voices": voices, "total": len(voices)}


@router.post(
    "/tts",
    response_class=Response,
    dependencies=[Depends(rate_limit)]
)
async def text_to_speech(
    text: Annotated[str, Body(embed=True, min_length=1, max_length=5000)],
    voice: Annotated[str, Body(embed=True)] = "Rachel",
    optimize_streaming: Annotated[bool, Body(embed=True)] = False,
    email: str = Depends(get_current_user)
):
    """
    Convert text to speech audio.
    
    Returns MP3 audio data.
    """
    audio_data = await audio_service.text_to_speech(
        text=text,
        voice=voice,
        optimize_streaming=optimize_streaming
    )
    
    return Response(
        content=audio_data,
        media_type="audio/mpeg",
        headers={
            "Content-Disposition": "inline; filename=speech.mp3",
            "Cache-Control": "public, max-age=3600"
        }
    )


@router.post(
    "/tts/stream",
    response_class=StreamingResponse,
    dependencies=[Depends(rate_limit)]
)
async def text_to_speech_stream(
    text: Annotated[str, Body(embed=True, min_length=1, max_length=10000)],
    voice: Annotated[str, Body(embed=True)] = "Rachel",
    email: str = Depends(get_current_user)
):
    """
    Stream text-to-speech audio for real-time playback.
    
    Useful for long content like articles or search result narration.
    """
    return StreamingResponse(
        audio_service.text_to_speech_stream(text=text, voice=voice),
        media_type="audio/mpeg",
        headers={
            "Content-Disposition": "inline; filename=speech_stream.mp3",
            "Cache-Control": "no-cache"
        }
    )


@router.post("/stt", dependencies=[Depends(rate_limit)])
async def speech_to_text(
    audio: UploadFile = File(..., description="Audio file (MP3, WAV, etc.)"),
    language: str = Form("en", description="Language code"),
    email: str = Depends(get_current_user)
):
    """
    Transcribe audio to text (speech-to-text).
    
    Supports voice search and audio transcription.
    """
    if not audio.content_type or not audio.content_type.startswith("audio/"):
        raise HTTPException(
            status_code=400,
            detail="Invalid file type - must be audio"
        )
    
    # Read audio file
    audio_content = await audio.read()
    
    # Create file-like object
    from io import BytesIO
    audio_file = BytesIO(audio_content)
    audio_file.name = audio.filename or "audio.mp3"
    
    result = await audio_service.speech_to_text(
        audio_file=audio_file,
        language=language
    )
    
    return result


@router.post(
    "/narrate/search",
    response_class=Response,
    dependencies=[Depends(rate_limit)]
)
async def narrate_search_results(
    query: Annotated[str, Body(embed=True)],
    results: Annotated[list[dict], Body(embed=True)],
    max_results: Annotated[int, Body(embed=True)] = 5,
    voice: Annotated[str, Body(embed=True)] = "Rachel",
    email: str = Depends(get_current_user)
):
    """
    Generate audio narration of search results.
    
    Useful for accessibility and hands-free search.
    """
    # Format results into narration text
    narration_text = audio_service.format_search_results_for_narration(
        query=query,
        results=results,
        max_results=max_results
    )
    
    # Generate audio
    audio_data = await audio_service.text_to_speech(
        text=narration_text,
        voice=voice,
        optimize_streaming=True
    )
    
    return Response(
        content=audio_data,
        media_type="audio/mpeg",
        headers={
            "Content-Disposition": f'inline; filename="search_narration.mp3"',
            "Cache-Control": "public, max-age=1800"
        }
    )


@router.post(
    "/notification",
    response_class=Response,
    dependencies=[Depends(rate_limit)]
)
async def audio_notification(
    message: Annotated[str, Body(embed=True, min_length=1, max_length=500)],
    notification_type: Annotated[str, Body(embed=True)] = "info",
    voice: Annotated[str, Body(embed=True)] = "Rachel",
    email: str = Depends(get_current_user)
):
    """
    Generate audio notification.
    
    Types: info, success, warning, error
    """
    valid_types = ["info", "success", "warning", "error"]
    if notification_type not in valid_types:
        raise HTTPException(
            status_code=400,
            detail=f"Invalid notification type. Must be one of: {valid_types}"
        )
    
    # Format notification
    notification_text = audio_service.format_notification(
        message=message,
        notification_type=notification_type
    )
    
    # Generate audio
    audio_data = await audio_service.text_to_speech(
        text=notification_text,
        voice=voice,
        optimize_streaming=True
    )
    
    return Response(
        content=audio_data,
        media_type="audio/mpeg",
        headers={
            "Content-Disposition": "inline; filename=notification.mp3",
            "Cache-Control": "no-cache"
        }
    )


@router.post(
    "/sound-effect",
    response_class=Response,
    dependencies=[Depends(rate_limit)]
)
async def generate_sound_effect(
    description: Annotated[str, Body(embed=True, min_length=1, max_length=200)],
    duration: Annotated[float, Body(embed=True, ge=0.5, le=10.0)] = 2.0,
    email: str = Depends(get_current_user)
):
    """
    Generate custom sound effect from description.
    
    Examples:
    - "notification bell"
    - "success chime"
    - "error buzzer"
    - "page turn"
    """
    audio_data = await audio_service.generate_sound_effect(
        description=description,
        duration_seconds=duration
    )
    
    return Response(
        content=audio_data,
        media_type="audio/mpeg",
        headers={
            "Content-Disposition": f'inline; filename="sound_effect.mp3"',
            "Cache-Control": "public, max-age=86400"  # Cache for 24h
        }
    )


@router.post(
    "/podcast/generate",
    response_class=Response,
    dependencies=[Depends(rate_limit)]
)
async def generate_podcast(
    title: Annotated[str, Body(embed=True)],
    content: Annotated[str, Body(embed=True, min_length=100, max_length=50000)],
    voice: Annotated[str, Body(embed=True)] = "Rachel",
    add_intro: Annotated[bool, Body(embed=True)] = True,
    email: str = Depends(get_current_user)
):
    """
    Generate podcast-style audio from long-form content.
    
    Useful for converting articles, documents, or search results
    into podcast episodes.
    """
    # Add intro if requested
    if add_intro:
        intro = f"Welcome to Nebula Search Podcast. Today's topic: {title}. "
        full_content = intro + content
    else:
        full_content = content
    
    # Generate audio with streaming optimization
    audio_data = await audio_service.text_to_speech(
        text=full_content,
        voice=voice,
        optimize_streaming=True
    )
    
    # Clean filename
    safe_title = "".join(c for c in title if c.isalnum() or c in (' ', '-', '_'))
    safe_title = safe_title.replace(' ', '_')[:50]
    
    return Response(
        content=audio_data,
        media_type="audio/mpeg",
        headers={
            "Content-Disposition": f'attachment; filename="podcast_{safe_title}.mp3"',
            "Cache-Control": "public, max-age=3600"
        }
    )

File: frontend/src/api/audio.js
/**
 * Audio API Client
 * Interface for ElevenLabs audio features
 */

const API_BASE = import.meta.env.VITE_API_URL || 'http://localhost:8000';

/**
 * Get authentication token from localStorage
 */
function getAuthToken() {
  return localStorage.getItem('token');
}

/**
 * Common fetch with auth headers
 */
async function authFetch(url, options = {}) {
  const token = getAuthToken();
  const headers = {
    ...options.headers,
    'Authorization': token ? `Bearer ${token}` : '',
  };

  const response = await fetch(url, { ...options, headers });
  
  if (!response.ok) {
    const error = await response.json().catch(() => ({ detail: 'Unknown error' }));
    throw new Error(error.detail || `HTTP ${response.status}`);
  }
  
  return response;
}

/**
 * Check if audio service is available
 */
export async function checkAudioHealth() {
  const response = await authFetch(`${API_BASE}/api/v1/audio/health`);
  return response.json();
}

/**
 * Get list of available voices
 */
export async function listVoices() {
  const response = await authFetch(`${API_BASE}/api/v1/audio/voices`);
  return response.json();
}

/**
 * Convert text to speech
 * @param {string} text - Text to convert
 * @param {string} voice - Voice name (default: "Rachel")
 * @param {boolean} optimizeStreaming - Enable streaming optimization
 * @returns {Promise<Blob>} Audio blob
 */
export async function textToSpeech(text, voice = 'Rachel', optimizeStreaming = false) {
  const response = await authFetch(`${API_BASE}/api/v1/audio/tts`, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    body: JSON.stringify({ text, voice, optimize_streaming: optimizeStreaming }),
  });
  
  return response.blob();
}

/**
 * Convert text to speech with streaming
 * @param {string} text - Text to convert
 * @param {string} voice - Voice name
 * @returns {Promise<Blob>} Audio blob
 */
export async function textToSpeechStream(text, voice = 'Rachel') {
  const response = await authFetch(`${API_BASE}/api/v1/audio/tts/stream`, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    body: JSON.stringify({ text, voice }),
  });
  
  return response.blob();
}

/**
 * Convert speech to text (transcribe audio)
 * @param {File|Blob} audioFile - Audio file to transcribe
 * @param {string} language - Language code (default: "en")
 * @returns {Promise<Object>} Transcription result
 */
export async function speechToText(audioFile, language = 'en') {
  const formData = new FormData();
  formData.append('audio', audioFile);
  formData.append('language', language);
  
  const token = getAuthToken();
  const response = await fetch(`${API_BASE}/api/v1/audio/stt`, {
    method: 'POST',
    headers: {
      'Authorization': token ? `Bearer ${token}` : '',
    },
    body: formData,
  });
  
  if (!response.ok) {
    const error = await response.json().catch(() => ({ detail: 'Transcription failed' }));
    throw new Error(error.detail);
  }
  
  return response.json();
}

/**
 * Generate audio narration of search results
 * @param {string} query - Search query
 * @param {Array} results - Search results array
 * @param {number} maxResults - Max results to narrate (default: 5)
 * @param {string} voice - Voice name
 * @returns {Promise<Blob>} Audio blob
 */
export async function narrateSearchResults(query, results, maxResults = 5, voice = 'Rachel') {
  const response = await authFetch(`${API_BASE}/api/v1/audio/narrate/search`, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    body: JSON.stringify({ query, results, max_results: maxResults, voice }),
  });
  
  return response.blob();
}

/**
 * Generate audio notification
 * @param {string} message - Notification message
 * @param {string} type - Notification type (info, success, warning, error)
 * @param {string} voice - Voice name
 * @returns {Promise<Blob>} Audio blob
 */
export async function audioNotification(message, type = 'info', voice = 'Rachel') {
  const response = await authFetch(`${API_BASE}/api/v1/audio/notification`, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    body: JSON.stringify({ message, notification_type: type, voice }),
  });
  
  return response.blob();
}

/**
 * Generate custom sound effect
 * @param {string} description - Sound description
 * @param {number} duration - Duration in seconds (0.5-10.0)
 * @returns {Promise<Blob>} Audio blob
 */
export async function generateSoundEffect(description, duration = 2.0) {
  const response = await authFetch(`${API_BASE}/api/v1/audio/sound-effect`, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    body: JSON.stringify({ description, duration }),
  });
  
  return response.blob();
}

/**
 * Generate podcast from content
 * @param {string} title - Podcast title
 * @param {string} content - Podcast content
 * @param {string} voice - Voice name
 * @param {boolean} addIntro - Add intro (default: true)
 * @returns {Promise<Blob>} Audio blob
 */
export async function generatePodcast(title, content, voice = 'Rachel', addIntro = true) {
  const response = await authFetch(`${API_BASE}/api/v1/audio/podcast/generate`, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    body: JSON.stringify({ title, content, voice, add_intro: addIntro }),
  });
  
  return response.blob();
}

/**
 * Play audio blob
 * @param {Blob} audioBlob - Audio blob to play
 * @returns {HTMLAudioElement} Audio element
 */
export function playAudio(audioBlob) {
  const audioUrl = URL.createObjectURL(audioBlob);
  const audio = new Audio(audioUrl);
  
  // Clean up URL when audio finishes
  audio.addEventListener('ended', () => {
    URL.revokeObjectURL(audioUrl);
  });
  
  audio.play();
  return audio;
}

/**
 * Download audio blob as file
 * @param {Blob} audioBlob - Audio blob to download
 * @param {string} filename - Filename for download
 */
export function downloadAudio(audioBlob, filename = 'audio.mp3') {
  const url = URL.createObjectURL(audioBlob);
  const a = document.createElement('a');
  a.href = url;
  a.download = filename;
  document.body.appendChild(a);
  a.click();
  document.body.removeChild(a);
  URL.revokeObjectURL(url);
}
